src/github.com/getlantern/flashlight/http/http.go

```diff
@@ -0,0 +1,167 @@
+package http
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/getlantern/flashlight/client"
+	"github.com/getlantern/flashlight/config"
+	"github.com/getlantern/flashlight/util"
+	"github.com/getlantern/golog"
+	"github.com/getlantern/whitelist"
+	"github.com/skratchdot/open-golang/open"
+	"net/http"
+	"reflect"
+	"time"
+
+	"github.com/getlantern/tarfs"
+)
+
+const (
+	UiUrl = "http://%s"
+	UiDir = "src/github.com/getlantern/ui/app"
+)
+
+var (
+	log = golog.LoggerFor("http")
+)
+
+type JsonResponse struct {
+	Error     string   `json:"Error, omitempty"`
+	Whitelist []string `json:"Whitelist, omitempty"`
+	Original  []string `json:"Original, omitempty"`
+}
+
+type WhitelistHandler struct {
+	http.HandlerFunc
+	whitelist *whitelist.Whitelist
+	wlChan    chan *whitelist.Config
+}
+
+func sendJsonResponse(w http.ResponseWriter, response *JsonResponse, indent bool) {
+
+	enc := json.NewEncoder(w)
+	err := enc.Encode(response)
+	if err != nil {
+		log.Errorf("error sending json response %v", err)
+	}
+}
+
+func setResponseHeaders(w http.ResponseWriter) {
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	w.Header().Set("Access-Control-Allow-Origin", "*")
+	w.Header().Set("Access-Control-Allow-Headers", "Content-Type, Content-Length, Accept-Encoding, X-CSRF-Token")
+	w.Header().Set("Access-Control-Allow-Credentials", "True")
+	w.Header().Set("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS")
+
+}
+
+func (wlh WhitelistHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	var response JsonResponse
+
+	setResponseHeaders(w)
+
+	switch r.Method {
+	case "OPTIONS":
+		// return if it's a preflight OPTIONS request
+		// this is mainly for testing the UI when it's running on
+		// a separate port
+		return
+	case "POST":
+		// update whitelist
+		decoder := json.NewDecoder(r.Body)
+		var entries []string
+		err := decoder.Decode(&entries)
+		if err != nil {
+			log.Error(err)
+			response.Error = fmt.Sprintf("Error decoding whitelist: %q", err)
+		} else {
+			wl := wlh.whitelist.UpdateEntries(entries)
+			copy := wlh.whitelist.Copy()
+			log.Debug("Propagating whitelist changes..")
+			wlh.wlChan <- copy
+			response.Whitelist = wl
+		}
+	case "GET":
+		response.Whitelist = wlh.whitelist.GetEntries()
+	default:
+		log.Debugf("Received %s", response.Error)
+		response.Error = "Invalid whitelist HTTP request"
+		response.Whitelist = nil
+	}
+	sendJsonResponse(w, &response, false)
+}
```

Comment:
As long as the websocket implementation handles reconnecting automatically,
I'm cool with it.

Cheers,
Ox

-----------------------------------------------------------------------------------------

"I love people who harness themselves, an ox to a heavy cart,
who pull like water buffalo, with massive patience,
who strain in the mud and the muck to move things forward,
who do what has to be done, again and again."

- Marge Piercy


On Sun, Feb 15, 2015 at 2:42 PM, atavism <notifications@github.com> wrote:

> In src/github.com/getlantern/flashlight/http/http.go
> <https://github.com/getlantern/flashlight-build/pull/41#discussion_r24725779>
> :
>
> > +		} else {
> > +			wl := wlh.whitelist.UpdateEntries(entries)
> > +			copy := wlh.whitelist.Copy()
> > +			log.Debug("Propagating whitelist changes..")
> > +			wlh.wlChan <- copy
> > +			response.Whitelist = wl
> > +		}
> > +	case "GET":
> > +		response.Whitelist = wlh.whitelist.GetEntries()
> > +	default:
> > +		log.Debugf("Received %s", response.Error)
> > +		response.Error = "Invalid whitelist HTTP request"
> > +		response.Whitelist = nil
> > +	}
> > +	sendJsonResponse(w, &response, false)
> > +}
>
> I thought we were. I actually have a websocket implementation working
> already so if there are no objections I think we should just go with that?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/getlantern/flashlight-build/pull/41/files#r24725779>.
>